CRM/Contribute/Form/ContributionRecur.php

```diff
@@ -0,0 +1,537 @@
+
+  /**
+   * The id of the contact associated with this recurring contribution record.
+   *
+   * @var int
+   */
+  public $_contactID;
+
+  /**
+   * form defaults
+   * @todo can we define this a as protected? can we define higher up the chain
+   * @var array
+   */
+  public $_defaults;
+
+  /**
+   * Values of existing recurring contribution record
+   */
+  public $_values;
+
+  /**
+   * The id of the related membership
+   */
+  public $_membershipID;
+
+  /**
+   * build all the data structures needed to build the form
+   *
+   * @return void
+   * @access public
+   */
+  function preProcess() { 
+    // Check permission for action.
+    if (!CRM_Core_Permission::checkActionPermission('CiviContribute', $this->_action)) {
+      CRM_Core_Error::fatal(ts('You do not have permission to access this page.'));
+    }
+    
+    $this->setPageTitle(ts('Recurring Contribution record'));
+
+    $this->_cdType = CRM_Utils_Array::value('type', $_GET);
+
+    $this->assign('cdType', FALSE);
+    if ($this->_cdType) {
+      $this->assign('cdType', TRUE);
+      CRM_Custom_Form_CustomData::preProcess($this);
+      return;
+    }
+
+    // Get the contact id
+    $this->_contactID = CRM_Utils_Request::retrieve('cid', 'Positive', $this);
+    $this->assign('contactID', $this->_contactID);
+
+    // Get the action.
+    $this->_action = CRM_Utils_Request::retrieve('action', 'String', $this, FALSE, 'add');
+    $this->assign('action', $this->_action);
+
+    // Get the contribution recur id if update
+    $this->_id = CRM_Utils_Request::retrieve('crid', 'Positive', $this);
+    if (!empty($this->_id)) {
+      $this->assign('contribRecurID', $this->_id);
+    }
+
+    $this->_context = CRM_Utils_Request::retrieve('context', 'String', $this);
+    $this->assign('context', $this->_context);
+
+    if ($this->_contactID) {
+      list($this->userDisplayName, $this->userEmail) = CRM_Contact_BAO_Contact_Location::getEmailDetails($this->_contactID);
+      $this->assign('displayName', $this->userDisplayName);
+    }
+    
+    if ($this->_action & CRM_Core_Action::DELETE) {
+      return;
+    }
+
+    $this->_values = array();
+
+    $ids = array();
+    $params = array('id' => $this->_id);
+		if (!empty($this->_id)) {
+			$recurring = new CRM_Contribute_BAO_ContributionRecur();
+      $recurring->copyValues($params);
+      $recurring->find(TRUE);
+      $ids['contributionrecur'] = $recurring->id;
+      CRM_Core_DAO::storeValues($recurring, $this->_values);
+
+      $membership = new CRM_Member_DAO_Membership();
+      $membership->contribution_recur_id = $this->_id;
+      $membership->is_test = 0;
+      if ($membership->find(true)) {
+        $this->_membershipID = $membership->id;
+        $this->assign('membershipID', $this->_membershipID);
+      }
+
+      // Get all backoffice payment processors
+      $backOfficePaymentProcessors = CRM_Contribute_BAO_ContributionRecur::getBackOfficePaymentProcessors();
```

Comment:
If you extend abstract payment & call the function referred to previously like the other forms then $this->_recurProcessors should be set already here